src/lib/updating-element.ts

```diff
@@ -0,0 +1,418 @@
+    this.root = this.attachShadow({mode : 'open'});
+  }
+
+  /**
+   * Calls `enableUpdate` and uses ShadyCSS to keep element DOM updated.
+   */
+  connectedCallback() {
+    if (window.ShadyCSS && this._hasUpdated) {
+      window.ShadyCSS.styleElement(this);
+    }
+    this.enableUpdate();
+  }
+
+  /**
+   * Call to enable element updating. This is called in `connectedCallback`
+   * but can be called sooner if desired.
+   */
+  enableUpdate() {
+    if (!this._isEnabled) {
+      // replay instance properties
+      for (const p in this._instanceProperties) {
+        // tslint:disable-next-line no-any
+        (this as any)[p] = this._instanceProperties[p];
+      }
+      this._isEnabled = true;
+      this.invalidate();
+    }
+  }
+
+  /**
+   * Synchronizes property values when attributes change.
+   */
+  attributeChangedCallback(name: string, old: string, value: string) {
+    if (old !== value) {
+      const ctor = (this.constructor as typeof UpdatingElement);
+      const propName = ctor._attributeToPropertyMap[name];
+      // TODO(sorvell): currently not used.
+      if (!this._serializingProperty) {
+        // tslint:disable-next-line no-any
+        (this as any)[propName] = ctor.propertyValueFromAttribute(propName, value);
+      }
+    }
+  }
+
+  /**
+   * Returns the value of the property with `name`. This method is used
+   * to get property values of registered properties.
+   */
+  protected getProperty(name: string) {
+    return this.props[name];
+  }
+
+  /**
+   * Sets the property `name` to the given `value`. This method is used
+   * to set property values of registered properties.
+   */
+  // tslint:disable-next-line no-any
+  protected setProperty(name: string, value: any) {
+    const old = this.props[name];
+    const ctor = (this.constructor as typeof UpdatingElement);
+    if (ctor.propertyShouldChange(name, value, old)) {
+      this.props[name] = value;
+      this.changedProps[name] = value;
+      const attrValue = ctor.propertyValueToAttribute(name, value);
+      if (attrValue !== undefined) {
+        // TODO(sorvell): doing this here means we need to allow round tripping
+        // through property setting to not interfere with the reaction stack.
+        // However, we could move this to update (as in Polymer) to avoid this
+        // with a flag, but this is undesirable since we want to keep update
+        // unimplemented here.
+        this.setAttribute(name, attrValue);
+      }
+      this.invalidate();
+    }
+  }
+
+  /**
+   * Call to request the element to asynchronously update regardless
+   * of whether or not any property changes are pending. This method is
+   * automatically called when any registered property changes.
+   */
+  protected async invalidate() {
+    if (this._isEnabled) {
+      if (this._isUpdating) {
+        console.warn('Requested an update while updating. This is not supported.');
+      } else if (!this._isValidating) {
+        this._isValidating = true;
```

Comment:
Seeing the use here, I think the name `_isValidating` is a little misleading. It's not that we're now validating, it's that we will validate, or are invalid. I think I'd prefer `_isInvalid` or `_isValid` if we don't want an inverted boolean.